src/core/qgstemporalnavigationobject.cpp

```diff
@@ -312,11 +312,14 @@ QgsTemporalNavigationObject::AnimationState QgsTemporalNavigationObject::animati
   return mPlayBackMode;
 }
 
-long QgsTemporalNavigationObject::findBestFrameNumberForFrameStart( const QDateTime &frameStart ) const
+long long QgsTemporalNavigationObject::findBestFrameNumberForFrameStart( const QDateTime &frameStart ) const
 {
-  long bestFrame = 0;
-  QgsDateTimeRange testFrame = QgsDateTimeRange( frameStart, frameStart ); // creatng an 'instant' Range here
-  for ( long i = 0; i < totalFrameCount(); ++i )
+  long long bestFrame = 0;
+  QgsDateTimeRange testFrame = QgsDateTimeRange( frameStart, frameStart ); // creating an 'instant' Range here
+  // earlier we looped from frame 0 till totalFrameCount() here, but this loop grew gigantic if you switched to for example milliseconds
+  // that is why now we calculate a rough framestart here
+  long long roughFrameStart = std::floor( ( frameStart - mTemporalExtents.begin() ).seconds() / mFrameDuration.seconds() );
```

Comment:
I'm still unsure about this logic -- it works fine under the assumption that the frame duration is fixed, but that's not the case anymore, and the actual frame duration is fluid (eg if it's set to intervals of months or years).

So if our frame duration was 1 month, then using mFrameDuration.seconds() will give a smaller duration then the actual frame duration for a 31 day month. (because the seconds will be calculated based on a 30 day month). In this case the roughFrameStart will be an overestimate, and we'll loop through all the remaining frames and never find the one we're after...

I think you'd need to base this on a worst-case scenario calculation, which I would be the 30/31 factor resulting from the 30 day month estimate vs potential 31 day actual month length. I.e. multiply the result by 0.9677419 before taking the floor of it.

But my maths may be wrong here :laughing:

Reply:
That is strange, I added a comment here before resolving the conservation... I added:

I think you are right, and to be honest I had the same feeling ! But I could not come up with some example.

See below for more explanation.